### DISCUSSION TITLE
Integration for Vizgen data

### DISCUSSION THREAD

**Discussion by yous0614** [giotto-suite/Giotto#457] [Q&A] [answered]
Hi team,

I am currently using GiottoSuite and I found the package smooth and amazing, I really like it.
However, I encounter a problem for integration of a couple of Vizgen datasets.
My goal is to integrate 10 human brain slides and at least to identify neurons, microglia, astrocytes cell types etc.

I was able to run the whole [vizgen brain] tutorial, however, the integration for the vizgen datasets isn't provided.
I then try to combine them using joinGiottoObjects() as in [Visium Prostate Integration] tag.
I was able to create a great align panel like this.
<img width="1336" alt="image" src="https://user-images.githubusercontent.com/57735418/205122901-f1f2a84a-fa63-48c5-834e-f0282107f336.png">

However, when I try to move on to filter or normalizeGiotto(), I found the error message like:

```
Error in matrixStats::colSums2(temp_matrix) :
  Argument 'x' must be of type logical, integer or numeric, not 'list'.
```
  
I wonder is there any solution for combine and analyse these dataset?

Shih-Feng
PhD student @ washU

**Comment by RubD** [comment]
Hi @yous0614 , can you provide a bit more detail? What's the exact script that you ran? It seems like the joint object doesn't have a raw count matrix. Did you use `calculateOverlapRaster` to identify which transcripts are within which cells (segmentations) and `overlapToMatrix` to aggregate the overlap information into a matrix?

**Reply by yous0614** [reply]
Thanks for the speedy replies.
Yes, so I followed mostly the tutorial on this page to prepare my giotto object. (https://giottosuite.readthedocs.io/en/latest/subsections/datasets/vizgen_mouse_brain.html),
I created the objects by `createGiottoObject()` and `addCellMetadata()`, then attached and update figure info on all the staining images by `createGiottoLargeImageList()`, `addGiottoImage()` and `updateGiottoLargeImage()`. Filtered, normalized, PCA, UMAP, NearestNetwork, LeidenCluster to generate the figure for each object. 

The tutorial doesn't mentioned how to integrate, I then run a for loop to make 10 objects in a list and join them by another tutorial on (https://giottosuite.readthedocs.io/en/latest/subsections/datasets/visium_prostate_integration.html) using:
```
vizgen_integrated = joinGiottoObjects(gobject_list = vizgen_objects,
                                      gobject_names = gobject_names, x_shift = c(seq(0, 15000*9, 15000)),
                                      join_method = 'shift', x_padding = 1000)
```
That's how I made the graph above, however, as you mentioned calculateOverlapRaster and overlapToMatrix, I went back and check vizgen@spatial_info, I realized I have to add the polygon and gpoint onto the file as the 'subcellular' section in the first tutorial.

`createGiottoPoints()` and `addGiottoPoints()` was used to make point data and attach.
`readPolygonFilesVizgen()` can't be used since  `H5Fopen_flags = "H5F_ACC_RDONLY"`  isn't an option, I used `readPolygonFilesVizgenHDF5()` then `addGiottoPolygons()` and `addSpatialCentroidLocations()` to attach the polygons.

By then, I can pass through `calculateOverlapRaster()` and `overlapToMatrix()`, however, I realized the spatial info in Vizgen object is stored in $z0, $z1, $z2, $z3, ...$z6, and though Slot "name", "spatVector", "spatVectorCentroids" are updated in every layer, the "overlaps" only updated in $z0, so I have to do `poly_info= 'z0'` to use it.

However, I still got 
```
> vizgen_integrated = normalizeGiotto(vizgen_integrated)
Error in matrixStats::colSums2(temp_matrix) :
  Argument 'x' must be of type logical, integer or numeric, not 'list'.
```
10 samples is too big so I tried to merge the first sample itself as test drive.
Here I attached the `showGiottoExpression()`, `showGiottoSpatialInfo()` and `showGiottoCellMetadata()` for before (a giotto object name vizgen) and after (merged object named vizgen_integrated) joinning object. 
Hope that can help. Thank you.

[vizgen and vizgen_integrated.txt](https://github.com/drieslab/Giotto/files/10148008/vizgen.and.vizgen_integrated.txt)

**Reply by yous0614** [reply]
Update: 
I figured perhaps the z0-z6 could be the problem, I tried to attach only one z-stack by `readPolygonFilesVizgenHDF5()`, along with gpoint data, I can perform `calculateOverlapRaster()` and `overlapToMatrix(..., poly_info = 'z3')`. By specify that, I can perform normalization on combined join Giotto object. But it warns:

```
> normalizeGiotto(combo1, spat_unit= 'z3', feat_type= 'rna')

 first scale feats and then cells
An object of class giotto
Steps and parameters used:

$`0_normalize`
          gobject         spat_unit         feat_type expression_values
         "combo1"              "z3"             "rna"             "raw"
     norm_methods library_size_norm       scalefactor          log_norm
       "standard"            "TRUE"            "6000"            "TRUE"
       log_offset           logbase       scale_feats       scale_cells
              "1"               "2"            "TRUE"            "TRUE"
      scale_order             theta       update_slot           verbose
    "first_feats"             "100"          "scaled"            "TRUE"

Warning message:
In libNorm_giotto(mymatrix = raw_expr[], scalefactor = scalefactor) :
  Total library size or counts for individual spat units are 0.
 This will likely result in normalization problems.
 filter (filterGiotto) or impute (imputeGiotto) spatial units.
```
I guess there should be a way to make a bypass by attaching HDF5 polygons as default in Slot "cell" but not "z3"?
Best,
Shih-Feng

**Comment by RubD** [comment]
Hi Shih-Feng, we're currently behind on our tutorials and examples, but we already have a working solution for dealing with multiple stacks, see the `aggregateStacks` function in suite or suite_dev (dev branch). We hope to update some of our tutorials to showcase it's usage.

**Reply by yous0614** [reply]
Hi Dr. Dries, 
thanks to point that out. I am not very familiar with the giotto object structure, so I tried to implement it by following one of the tutorial I found. (https://github.com/drieslab/GiottoData/blob/b22ff768159b4173f29dadfecd86cb176593ca95/inst/Mini_datasets/Vizgen/mini_vizgen_script.R)

Here I tried to combine the vizgen object twice by inserting the `joinGiottoObjects()` after step 2 in different ways, I hope you can correct me that which way is more logical, or is there any parameter I set it wrong.

**1.** `joinGiottoObjects()` -> `calculateOverlapRaster()` -> `overlapToMatrix()` -> `aggregateStacks()` ->  `filterGiotto()` -> `normalizeGiotto()`
```
testcombo1 = joinGiottoObjects(c(vizgen, vizgen), c('v1', 'v2'), join_method= 'shift', x_shift = c(0, 15000), x_padding= 1000)
testcombo1 = calculateOverlapRaster(testcombo1, spatial_info= 'z0', feat_info= 'rna', feat_subset_column= 'global_z', feat_subset_ids= 0)
testcombo1 = overlapToMatrix(testcombo1, poly_info = 'z0', feat_info = 'rna', name = 'raw')
testcombo1 = calculateOverlapRaster(testcombo1, spatial_info= 'z1', feat_info= 'rna', feat_subset_column= 'global_z', feat_subset_ids= 1)
testcombo1 = overlapToMatrix(testcombo1, poly_info = 'z1', feat_info = 'rna', name = 'raw')
testcombo1 = aggregateStacks(testcombo1, spat_units= c('z0', 'z1'), feat_type= 'rna', values= 'raw', summarize_expression= 'sum', summarize_locations= 'mean', new_spat_unit= 'aggregate')
testcombo1 = filterGiotto(testcombo1, spat_unit= 'aggregate', expression_threshold = 1, feat_det_in_min_cells = 3, min_det_feats_per_cell = 5, expression_values= 'raw', poly_info= 'z0')
```  
```
completed 1: preparation
Error in S4_expr@exprMat[filter_bool_feats, ] :
incorrect number of dimensions
```  
**2.**`calculateOverlapRaster()` -> `overlapToMatrix()` -> `aggregateStacks()`->`joinGiottoObjects()` -> `filterGiotto()` -> `normalizeGiotto()`
  
```
vizgen = calculateOverlapRaster(vizgen, spatial_info= 'z0', feat_info= 'rna', feat_subset_column= 'global_z', feat_subset_ids= 0)
vizgen = overlapToMatrix(vizgen, poly_info = 'z0', feat_info = 'rna', name = 'raw')
vizgen = calculateOverlapRaster(vizgen, spatial_info= 'z1', feat_info= 'rna', feat_subset_column= 'global_z', feat_subset_ids= 1)
vizgen = overlapToMatrix(vizgen, poly_info = 'z1', feat_info = 'rna', name = 'raw')
vizgen = aggregateStacks(vizgen, spat_units= c('z0', 'z1'), feat_type= 'rna', values= 'raw', summarize_expression= 'sum', summarize_locations= 'mean', new_spat_unit= 'aggregate')
testcombo2 = joinGiottoObjects(c(vizgen, vizgen), c('v1', 'v2'), join_method= 'shift', x_shift = c(0, 15000), x_padding= 1000)
```  
```
Join method:shift
start updating objects
A) Update giotto Objects
0. Update cell and feature IDs
1. Update expression IDs
Error in dimnamesGets(x, value) :
length of Dimnames[[2]] (32271) is not equal to Dim[2] (32688)

```    
**3.**`filterGiotto()` -> `joinGiottoObjects()` -> `calculateOverlapRaster()` -> `overlapToMatrix()` -> `aggregateStacks()` ->  `normalizeGiotto()`
```
vizgen = filterGiotto(gobject = vizgen, expression_threshold = 1, feat_det_in_min_cells = 100, min_det_feats_per_cell = 20)
testcombo3 = joinGiottoObjects(c(vizgen, vizgen), c('v1', 'v2'), join_method= 'shift', x_shift = c(0, 15000), x_padding= 1000)
testcombo3 = calculateOverlapRaster(testcombo3, spatial_info= 'z0', feat_info= 'rna', feat_subset_column= 'global_z', feat_subset_ids= 0)
testcombo3 = overlapToMatrix(testcombo3, poly_info = 'z0', feat_info = 'rna', name = 'raw')
testcombo3 = calculateOverlapRaster(testcombo3, spatial_info= 'z1', feat_info= 'rna', feat_subset_column= 'global_z', feat_subset_ids= 1)
testcombo3 = overlapToMatrix(testcombo3, poly_info = 'z1', feat_info = 'rna', name = 'raw')
testcombo3 = aggregateStacks(testcombo3, spat_units= c('z0', 'z1'), feat_type= 'rna', values= 'raw', summarize_expression= 'sum', summarize_locations= 'mean', new_spat_unit= 'aggregate')
testcombo3 = normalizeGiotto(testcombo3, spat_unit= 'aggregate', scalefactor= 5000, verbose=T)
```
```
Error in x$.self$finalize() : attempt to apply non-function
In addition: Warning message:
In libNorm_giotto(mymatrix = raw_expr[], scalefactor = scalefactor) :
Total library size or counts for individual spat units are 0.
This will likely result in normalization problems.
filter (filterGiotto) or impute (imputeGiotto) spatial units.
Error in x$.self$finalize() : attempt to apply non-function
  
first scale feats and then cells
Error in x$.self$finalize() : attempt to apply non-function
Error in x$.self$finalize() : attempt to apply non-function
```

**Comment by yous0614** [comment, marked as answer]
Update:

Thanks for the reply.
I found my problem, when use `joinGiottoObjects()`, in object@expression it generates not only a cell-by-gene `$matrix`, but also a list of genes called `$sort_all_feats`. The `$Matrix` and `$sort_all_feats` in `object@expression$cell$rna$raw` are considered a `list`, therefore it causes a problem when `get_expression_values()` trying to get a matrix from vizgen object and put forward in `filterGiotto()`, which causes:

```
Error in rowSums_flex(expr_values >= expression_threshold) :
  'list' object cannot be coerced to type 'double'
```
I bypassed the problem by 
```
vizgen_integrated@expression$cell$rna$raw = 
get_expression_values(vizgen_integrated, spat_unit= 'cell', feat_type= 'rna', output = 'matrix', values= 'raw')[[1]]
```

Thanks
